projects/concurrency/README.md

```diff
@@ -0,0 +1,308 @@
+
+Try to modify the buggy code above to use atomics. Make sure that if you run it a few times, you always end up with `1000`.
+
+##### Memory barriers / fences
+
+Under the covers, something involved in running our program (possibly the Go standard library, possibly something in the operating system, possibly even something in the CPU) is inserting something called [a memory barrier (also known as a fence)](https://en.wikipedia.org/wiki/Memory_barrier). This is an instruction to say "it's important that this operation is visible to other cores", which forces different cores to synchronise their views of a piece of memory before continuing.
+
+This can be slow. Recall that we have these per-core caches because they're so much faster than sharing memory across cores. Every time we need to insert one of these memory barriers, we slow down our program. This is why memory access isn't all atomic by default. But it can be worth it.
+
+#### When are atomics good/bad to use?
+
+Atomics can be really useful when you need to operate on a single item of small data (e.g. a number). They are simple and easy to reason about. But they typically operate on one item in memory. If you need to operate on larger data (e.g. a string), multiple pieces of memory (e.g. two different numbers which must be changed together), or perform an operation not supported by atomics (e.g. multiplication), you probably can't use an atomic for that.
+
+### Mutexes
+
+[The golangbot Mutex section](https://golangbot.com/mutex/#mutex) used a `Mutex` instead of an atomic to solve the same problem. Mutexes are more flexible - you choose when you lock and unlock them, but more complicated - you need to make sure you unlock them at the right place. They are also prone to deadlock: Blocking your programme from making progress because two threads are waiting on locks held by each other.
+
+If you have multiple operations which need to appear as if they were atomic (i.e. you must prevent other threads from taking action between the operations), a Mutex is often a good way to achieve this.
+
+#### Critical Sections
+
+When using a Mutex, it's crucial to work out where the Mutex needs to be acquired ("locked") and where it needs to be released ("unlocked").
+
+In general, any time that some other thread changing something between operations would cause a problem, those operations need to be performed _under the same lock acquisition_.
+
+Let's take an example of tracking some statistics. Imagine we have a counter for total number of requests, and counters for successful and failed requests. We may want the invariant to always hold: (successful requests + failed requests) = total requests.
+
+This code would be buggy:
+
+```go
+statsMutex.Lock()
+if success {
+    successfulRequests += 1
+} else {
+    failedRequests += 1
+}
+statsMutex.Unlock()
+statsMutex.Lock()
+totalRequests += 1
```

Comment:
it will eventually be correct, though. i feel like we could have a more compelling example here.
i will think about this over lunch!

Reply:
so i tried out an example with a double-linked list with a buggy split critical section but it's a bit too complex to be clear. will see if i can work out something better.

Reply:
How about something like the below. Because the stats object in the map is initialised in a different lock section to the initial check, we could have subsequent goroutines reinitialise the map value after previous goroutines have written stats, causing data loss. Also, it's a bit more plausible for people to not realise that they need to maintain a critical section (or redo the nil check) when initialising like this, it's a classic concurrency error.

```
type custStats struct {
	failedReqs  int
	successReqs int
}

var (
	lock            sync.Mutex
	statsByCustomer = make(map[string]*custStats)
)

func updateCustStats(custId string, ok bool) {
	// does customer stats object exist in map?
	lock.Lock()
	_, exists := statsByCustomer[custId]
	lock.Unlock()

	// create stats obj if necessary
	if !exists {
		lock.Lock()
		statsByCustomer[custId] = &custStats{}
		lock.Unlock()
	}

	lock.Lock()
	if ok {
		statsByCustomer[custId].successReqs++
	} else {
		statsByCustomer[custId].failedReqs++
	}
	lock.Unlock()
}
```